### PR TITLE
Activation Tuning Refactor

### PR DESCRIPTION
**Pull request by emersodb**
Migrating the activation tuning notebooks to use LLaMA-2 instead of OPT. There is less of a dramatic impact in terms of performance improvement when including prompts, but it's still there.

1. This refactor includes saving the activations across several layers of LLaMA-2 with and without prompts. I also moved the OPT-175 activations into their own folder. These changes represent the majority of "file changes" in this PR.
    **NOTE** It take a long time to compute these activations, so we're saving them for participant convenience.
3. Update to the readme to reflect the migration
4. Updates to the calculation and saving of activations
5. Updates to the training notebook on the activations themselves.

### CODE REVIEW

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb [hunk 1]
```diff
@@ -87,31 +87,505 @@
...
      },
-     "execution_count": 7,
+     "execution_count": 5,
      "metadata": {},
      "output_type": "execute_result"
     }
    ],
    "source": [
-    "model.module_names[-10:]"
+    "model.module_names"
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb:588]: Looks like we're listing all the modules instead of the last 10 as described above?

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb:588]: Yes, good call. I'll modify the comment.

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb [hunk 2]
```diff
@@ -131,40 +605,76 @@
...
      "text": [
-      "Activations(activations=[{'decoder.layers.95.fc2': tensor([[ 0.0062, -0.4624,  0.4702,  ...,  0.0985,  1.2588, -0.7002],\n",
-      "        [-0.2197,  0.3364, -0.1965,  ..., -0.2651,  0.0475, -0.1638]],\n",
-      "       dtype=torch.float16)}, {'decoder.layers.95.fc2': tensor([[ 2.5312, -1.4609, -1.7266,  ...,  1.9453, -0.8262, -1.5234],\n",
-      "        [-0.2294, -0.6396,  0.9043,  ..., -0.8140,  0.1183,  0.2437],\n",
-      "        [-0.8062,  0.8320, -0.8003,  ..., -0.8530, -0.7656,  0.5718]],\n",
-      "       dtype=torch.float16)}], logprobs=[[None, -7.15677547454834, -6.544065475463867], [None, -5.25247859954834, -6.673819065093994, -5.898566246032715]], text=['Hello World', 'Fizz Buzz'], tokens=[['</s>', 'Hello', ' World'], ['</s>', 'F', 'izz', ' Buzz']])\n",
-      "Tensor Shape: torch.Size([2, 12288])\n",
-      "Tensor Shape: torch.Size([3, 12288])\n"
+      "Activations(activations=[{'layers.0': tensor([[ 0.0464,  0.0161,  0.0542,  ...,  0.0118,  0.0105, -0.0007],\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb:610]: As discussed, I think it could be useful to add a comment saying that currently we only support retrieving the activations from a single layer 🙂

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb:610]: Will do!

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb [hunk 3]
```diff
@@ -371,7 +1064,7 @@
   },
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/compute_activations.ipynb:1066]: Typo: OPT -> Llama2

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb [hunk 4]
```diff
@@ -249,7 +276,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "We now load the activations associated with a small training set of 100 samples and a test set with 300 samples that were generated using prompts as part of the input to the OPT model. The prompt structure can be seen in the `compute_activations.ipynb` notebook, but they incorporate 5-shot examples and an instruction prompt."
+    "We now load the activations associated with a small training set of 100 samples and a test set with 300 samples that were generated using prompts as part of the input to the OPT model. The prompt structure can be seen in the `compute_activations.ipynb` notebook, but they incorporate few-shot examples and an instruction prompt."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb:279]: "OPT" -> "Llama-2"

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb [hunk 5]
```diff
@@ -258,26 +285,72 @@
...
+    "train_and_evaluate_model(model, train_dataloader, test_dataloader, device)"
    ]
   },
   {
    "attachments": {},
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "We can now write a relatively simple script to train and evaluate our model. The model has an embedding dimension of 12,288 corresponding to the size of the intermediate activations for OPT-175B. The hidden dimension is small (128) and the final dimension corresponds to our label space (positive, negative)."
+    "This is quite an amazing result. Simply by including a few-shot prompt when computing the activations, we have __significantly__ increased the sampling efficiency of training this small classifier and induced an immense jump in performance."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb]: I know there is definitely an improvement in performance, but just wondering if we want to characterize it as an immense jump (0.863 -> 0.8867)... 😂 ?

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb [hunk 6]
```diff
@@ -289,66 +362,135 @@
...
+   "metadata": {},
+   "source": [
+    "There are a few interesting takeaways from these results:\n",
     "\n",
-    "            logits = model(activations)\n",
-    "            predictions.extend((logits.argmax(dim=1) == labels))\n",
+    "* The prompt structures we provided helped to improve the results. For the best performing layer activations we move from `0.856%` accuracy without prompts to `0.893%` with them.\n",
     "\n",
-    "        accuracy = torch.stack(predictions).sum() / len(predictions)\n",
+    "* In both cases, the best layer activations for the task are not the first or last layers, but rather Layer 20. This has been observed in other contexts, as the earlier layers contain more general embeddings.\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb:480]: Is this also true for OPT? Namely, is the best performing layer also neither the first nor the last layer?

File: src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb (hunk 5 above)
Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb]: Very good point. OPT gets an immense jump in accuracy (hence the wording left over from the previous setup). This is much less substantial. I just noted that it was a "measurable" jump in accuracy.

**Review by yuchongzhang — APPROVED** [review]
Everything looks good to me! Thanks for addressing the comments.
